File: ruby/grade-school/school.rb
class School
  def initialize
    @students = Hash.new{|h, k| h[k] = []}
  end
  def add (v, k)
    @students[k] << v
    @students[k].sort!
  end
  def to_hash
    Hash[@students.sort]
  end
  def grade (k)
    @students[k]
  end
end

File: haskell/grade-school/School.hs
module School (empty, add, sorted, grade, School) where
import Data.IntMap (findWithDefault, toList, insertWith, IntMap)
import qualified Data.IntMap (empty)
import Data.List (sort)

newtype School = School (IntMap [String])

empty :: School
empty = School (Data.IntMap.empty)

add :: Int -> String -> School -> School
add grade name (School school) =
  School $ insertWith (\x y -> sort $ x++y) grade [name] school

sorted :: School -> [(Int, [String])]
sorted (School school) = toList school

grade :: Int -> School -> [String]
grade n (School school) = findWithDefault [] n school
